### ISSUE TITLE
Implementing Player Look (Yaw and Pitch)

### ISSUE THREAD

**Issue by repository** [ORelio/Minecraft-Console-Client#691]
Hello, I've been trying to implement a look command, but this issue has gotten me stumped.
I've been trying to update the player's yawpitch by setting the value `yawpitch` in  `McTcpClient.cs`, however, this does not affect anything. My code has been attached below. Thank you.
![code](https://user-images.githubusercontent.com/41586666/55530134-ab728d00-5659-11e9-9949-4204e65b2bc6.jpeg)

Edit: values yaw and pitch are both singles.

**Comment by Pokechu22** [comment]
You'd need to send a move packet afterwards.  I'm not 100% sure how you do it (it's been quite some time since I've worked with this project; probably just doing something that would normally trigger a move would work).  But that should be enough.

(if your `yaw` and `pitch` variables are single-precision floats, that should be correct, though you _might_ need to do something with endianity.  Again, it's been a while for me.)

**Comment by repository** [comment]
There's a function run 20 times a second that sends the player position and look packet.
https://github.com/ORelio/Minecraft-Console-Client/blob/master/MinecraftClient/McTcpClient.cs#L534
I haven't tried anything with endianity yet, though I'll fiddle with it and see if I can get it working.

Also, thanks for the fast reply, didn't expect it to be that fast.

**Comment by ORelio** [comment]
Hi,

You need to have terrain and movements enabled, which is currently not implemented for MC 1.13. However, you are on the right track so I'd advise testing on MC 1.12 or lower while implementing that feature.

The `SendLocationUpdate` method sends a [Player Position And Look](https://wiki.vg/Protocol#Player_Position_And_Look_.28serverbound.29) packet so that's what you want to do. Currently, player look is not implemented. Upon logging in, it is received from server, stored as is in `yawpitch` byte array, and sent back on each location update. To implement it properly, I'd advise implementing yaw and pitch decoding/encoding so that the method takes a yaw and a pitch, and perform the encoding job in the protocol handler.

This means the following prototypes would be used instead for `UpdateLocation` and `SendLocationUpdate`:

```C#
public void UpdateLocation(Location location, float yaw, float pitch);
SendLocationUpdate(location, Movement.IsOnGround(world, location), yaw, pitch);
```

... making alteration of yaw and pitch much easier in McTcpClient.

In order to do so, update the prototypes in `McTcpClient` and protocol handlers, implement `readNextFloat` and `getFloat` in the `Protocol18` protocol handler (see existing `readNextDouble` and `getDouble` functions, that's nearly copy, paste and write `Float` instead of `Double`), and adjust data decoding/encoding to actually decode the float values before calling `UpdateLocation`, then reencode them before sending back to the server inside the `SendLocationUpdate` method. Once yaw and pitch support is implemented, you can alter it more easily inside protocol handler, and scripts/bots if you add the relevant API methods.

For calculation of yaw and pitch, see [Player Look](https://wiki.vg/Protocol#Player_Look).

**Comment by ORelio** [comment]
Oh, now that I think about it, maybe it would make the calculation easier by adding a metho that takes two locations (player location and coordinates to look at) and returns appropriate yaw and pitch using the vector formula on the Wiki. This way, the ChatBot & Script API would have a method like this:

PlayerLookAt(Mapping.Location location)

...which would be easy and straightforward to use.

**Comment by ORelio** [comment]
Looks good! For debugging, I'd advise trying simple values such as 90 or -90 and see ingame if the head is moving properly. Also make sure to test on a vanilla server, as more realistic head movements might be necessary on servers with anti-cheat plugins. If still having issues, feel free to upload your code (clone this repo, then push your changes to the clone) so that I can have a look at it.

**Comment by repository** [comment]
Thanks, and I've been testing on a 1.12.2 vanilla server. I'll try out simple values and see where I get with that.

**Comment by repository** [comment]
Success! The player looks around, but not in the right direction... 

Perhaps there is something wrong with my formula. If so, I don't see it.

I've made a fork and pushed my code to it.

https://github.com/repository/Minecraft-Console-Client/blob/a4fdc89943374e7e80fee4883a0a201edb3c6480/MinecraftClient/McTcpClient.cs#L429

![image](https://user-images.githubusercontent.com/41586666/55844230-22dc6c80-5af1-11e9-8b45-d6a8d94d4be8.png)
I tried to make the bot look at the block it's standing on.
